Add autocomplete suggestions for tags and handles

mariodian opened the pull request:
solves #1569

morebrownies commented:
very nice 👍 👍 

Is it possible to reverse the order of the autocomplete so that the most recently typed handle or tag displays at the top of the list instead of the bottom? Currently the oldest handle and tag I typed display at the top and I'm afraid that list will become stale over time.

mariodian commented:
It'should be sorted alphabetically now (for easier reading). 

I can sort it in anyway you like though.

rmisio commented on js/views/pageNavVw.js:
```diff
@@ -617,19 +798,19 @@ module.exports = baseVw.extend({
 
-    this.ServerSubmenuTimeout = setTimeout(() => {
-      this.serverSubmenu.addClass('server-submenu-opened');
+    this.$serverSubmenuTimeout = setTimeout(() => {
```
I'm totally for prefacing cached jQuery elements with a $, but this is an integer representing a timeout (i.e. not a jQuery object), so I don't think we should be prefacing it with the dollar sign.

jjeffryes replied:
I agree. $ prefixes should be used only to designate cached jquery objects.

mariodian replied:
Thanks, I'll fix it.

rmisio commented on js/views/pageNavVw.js:
```diff
@@ -591,14 +611,175 @@ module.exports = baseVw.extend({
       Backbone.history.navigate(route, {trigger: true});
     });
   },
+  
```
Does this have to live in pageNavView? Could we not have a seperate suggestionListView (or something) that encapsulates the functionality to render a list of suggestions with a way to navigate through them? Then pageNavView would be the parent view interfacing with our separate child view.

I think in general we have overly bloated views and it would be more manageable to break isolated functionality into separate views.

I'm fine with not using a plug-in (not sure that I recommend it), but I'd prefer we encapsulate the functionality in a more decoupled and re-usable way. Say, for example, we want to add this suggestion menu to a text field in a different view (which has already been discussed). With it being so coupled to the page nav, we'd have a good re-factor on our hands.

Without using a plugin, you might be re-inventing the wheel to do it right.

jjeffryes replied:
The ideal would be to abstract this into an autosuggest view that can be re-used with any source of suggestions, so we could use it on other search fields.

rmisio replied:
If we found a lean plugin with a rich enough API for our needs, not sure why we would want to create this on our own.

mariodian replied:
Problem with such approach is that suggestionListView wouldn't be able to access the search input the backbone way - this.$('.element'). I've seen accessing various elements from non-related views though, so I guess we don't care about it that much? 

If you, on the other hand, put the input to suggestionsList.html then you'd have to move all the addressBar*() methods from pageNavView to the suggestionsListView and have a problem accessing closeStatusBar() that is called in addressBarKeyup(), and also called from within pageNavView numerous times.

At first, I made a separate view for suggestions and then I stumbled upon above issues and couldn't find a "pretty" way to solve it. That's why everything ended up in pageNavView. I know it isn't ideal.

rmisio replied:
You're right, views should not access each other's DOM elements. But, views can communicate in other ways depending on the situation, namely by calling each other's methods (if you have a reference to other view). Also, views can communicate via events.

If we were using a plug-in, it would probably be fine for this to be in this view, since the plug-in would encapsulate a lot of the common functionality. To be honest, if we don't use a plugin here, whoever is tasked with implementing this on a text field in a different view, will probably need to bring a plug-in in and then refactor this code.

FWIW, I'd recommend a plugin as long as we found a suitable one.

mariodian replied:
This was my iteration:

1) jQuery plugin (I had some issue with it, can't recall now what it was)
2) a separate view (issues mentioned above)
3) everything in pageNavView

I'll look at it tomorrow again whether it can end up in a plugin. But I think searching through saved data (tags and handles) couldn't be abstract enough because they have a different structure.

rmisio replied:
the part of searching through save data is our custom logic that wouldn't live in the plugin. A plugin should give you events / hooks where you could impart your own custom logic. A big thing we're looking for the plugin to give us is the common UI stuff, like keyboard navigation, which we wouldn't want to have to repeat for each area we want autosuggest.

rmisio commented on js/views/userPageVw.js:
```diff
@@ -307,6 +307,25 @@ module.exports = baseVw.extend({
+          
+          // Handle was requested
+          if (self.model.get('page').profile.handle) {
+            var profile = self.model.get('page').profile,
```
There already is a `profile` variable declared in this same scope, a few lines above. I think it might correspond to the same thing you need. If so, please use it, otherwise please rename your variable.

On a different note, I know we're not 100% consistent in this, but I think we should be making our variable declarations at the top of each scope, both to avoid situations like this and also to accurately mimic variable hoisting. @jjeffryes What do you think - are you fine with us adopting that convention (at least going forward)?

If you do indeed want block level scope, feel free to use ES6's `let`, which is supported as long as this module is in strict mode (ensure there is a `use strict;` declaration up top).

rmisio replied:
One mild concern I have about this is that we are continually adding to localStorage here. While the LS limit is pretty high (at least relative to cookies), it's still finite and we wouldn't want to run out of space. I'm wondering if we should cap the list at some number - could be high (maybe 1000). After that, for any new handles, we should remove the oldest stored handle first to create space for the new one. I suppose this would require you to store a timestamp with each handle (which, humorously enough, would take addition space to store).

@jjeffryes What do you think? Valid concern, or am I overthinking? We definitely don't want to fill up local storage because we use it for more critical things than this.

mariodian replied:
> There already is a profile variable declared in this same scope, a few lines above.

Thanks I haven't noticed that.

jjeffryes replied:
I don't have it set as a rule in es-lint, but generally variables should be set at the top, which would have avoided the collision here. ES-Lint should have picked up the duplicate variable if this is up to date with master (if it isn't, merge it and re-lint it).

jjeffryes replied:
I'm not too worried about running out of localStorage space, but we also don't want to invite trouble with potentially infinite lists of things. Capping it at 1000 would be fine.

rmisio replied:
To make it easy, forget the timestamp stuff I mentioned earlier. Just cap it at some number and then stop adding to the list. Maybe even 5000 would be ok.

mariodian replied:
I can always put the latest tag/handle at the end of the array and then reverse() it so we have a list of latest X tags/handles.

jjeffryes commented on js/views/userPageVw.js:
```diff
@@ -307,6 +307,25 @@ module.exports = baseVw.extend({
           !self.currentItemHash && self.loadingDeferred.resolve();
+          
+          // Handle was requested
+          if (self.model.get('page').profile.handle) {
```
Is the userPage view the right place to record handles, or would this be better to do in the router, which is already parsing handles?

mariodian replied:
AFAIK you can only tell whether the handle is valid (existent) in userPageView. I think we don't want to save invalid handles?

rmisio replied:
The router only processes handles if someone enters @<username> in the address bar.

So, the userPage is a more comprehensive place to get the handle, since it will get it whether you're accessing the page via @<username>, guid, etc... (and clicking on a username link usually accesses the user view via guid, so the router wouldn't pick that up).

The only thing I would consider changing is instead of storing the logic to record handles in the userPage, I would have the userPage fire an event when a handle is retrieved and then catch that event in pageNavView and record the handles there. The benefit of that is that you are centralizing all the functionality dealing with storing and retrieving the handles from LS in one place.

mariodian replied:
> I would have the userPage fire an event when a handle is retrieved and then catch that event in pageNavView and record the handles there.

Then it should be the same for tags that are currently handled in homeView.

rmisio replied:
tru dat

jjeffryes replied:
I suppose the intent is to auto-complete the handles of nodes you've visited, vs. auto-complete based on what you typed.

If we did it in the router, it would record handles you tried to reach that didn't resolve, which you might want to try again later. Having something you know you've typed not show up in the suggestions could look like a bug. 

On the other hand, it's useful to not suggest handles if we don't think they are valid, which will also prevent typos from becoming suggestions.

mariodian commented:
Based on objections above I changed the code:
- tags/handles history saving is handled by pageNavView (triggered via userPageView and HomeView respectively)
- added list pruning (currently max 1000 items)
- lists sorted from latest to oldest

I haven't changed the suggestion box because I feel we still haven't figured whether we want to use a plugin, make our own, or put it in a separate view.

rmisio commented on js/views/homeVw.js:
```diff
@@ -470,17 +470,21 @@ module.exports = baseVw.extend({
     if (searchItemsText){
+      var hashedItem = "#" + searchItemsText;
+      
+      window.obEventBus.trigger('saveVisitedTag', hashedItem);
```
Could we please make the event type a more generic indicator of what is happening, rather than a directive to do something? Maybe something like 'searchingText'? Other pieces of code may want to hook into the same event.

jjeffryes replied:
I agree. The event being fired is "searchItem just happened". It may get used to save the tag, but if some other code wants to know when searchItem occurs, we don't want to add a second event for the same thing.

mariodian replied:
Sure

rmisio commented on js/views/userPageVw.js:
```diff
@@ -307,6 +307,11 @@ module.exports = baseVw.extend({
+          
+          // Handle was requested
+          if (profile.handle) {
+            window.obEventBus.trigger('saveVisitedHandle', profile);
```
Same comment as above... can the event type be more descriptive of what is happening? Maybe 'handleObtained' ...?

mariodian committed:
Add autocomplete suggestions for tags and handles

jjeffryes commented on js/views/pageNavVw.js:
```diff
@@ -490,14 +567,20 @@ module.exports = baseVw.extend({
+    if (!$target.hasClass('js-navAddressBar') && this.suggestionsVisible) {
+      app.hideOverlay();
+      this.hideSuggestions();
+    } else if ($target.hasClass('js-navAddressBar') || !($target.hasClass('popMenu') ||
```
Removing comment, it was just funky formatting in my IDE
